### PR TITLE
[WIP] feat: add pause for MonoVertex

### PR DESCRIPTION
**Pull request by kohlisid**

### CODE REVIEW

File: pkg/apis/numaflow/v1alpha1/mono_vertex_types.go [hunk 1]
```diff
@@ -607,3 +625,43 @@ type MonoVertexList struct {
...
+	//// +optional
+	//DeleteGracePeriodSeconds *int32 `json:"deleteGracePeriodSeconds,omitempty" protobuf:"varint,1,opt,name=deleteGracePeriodSeconds"`
+	// DesiredPhase used to bring the pipeline from current phase to desired phase
+	// +kubebuilder:default=Running
+	// +optional
+	DesiredPhase MonoVertexPhase `json:"desiredPhase,omitempty" protobuf:"bytes,1,opt,name=desiredPhase"`
+	// PauseGracePeriodSeconds used to pause pipeline gracefully
+	// +kubebuilder:default=30
+	// +optional
+	PauseGracePeriodSeconds *int32 `json:"pauseGracePeriodSeconds,omitempty" protobuf:"varint,2,opt,name=pauseGracePeriodSeconds"`
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/mono_vertex_types.go:642]: Do not need this, we do not do anything for graceful pausing.

File: pkg/apis/numaflow/v1alpha1/mono_vertex_types.go [hunk 2]
```diff
@@ -607,3 +625,43 @@ type MonoVertexList struct {
 	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 	Items           []MonoVertex `json:"items" protobuf:"bytes,2,rep,name=items"`
 }
+
+type MonoVertexLifecycle struct {
+	// TODO(MonoVtx-Pause): Do we want to add delete grace period also?
+	//// DeleteGracePeriodSeconds used to delete a MonoVertex gracefully
+	//// +kubebuilder:default=30
+	//// +optional
+	//DeleteGracePeriodSeconds *int32 `json:"deleteGracePeriodSeconds,omitempty" protobuf:"varint,1,opt,name=deleteGracePeriodSeconds"`
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/mono_vertex_types.go:634]: No need, we do nothing for graceful deletion.

File: pkg/apis/numaflow/v1alpha1/mono_vertex_types.go [hunk 3]
```diff
@@ -38,6 +38,8 @@ const (
 	MonoVertexPhaseUnknown MonoVertexPhase = ""
 	MonoVertexPhaseRunning MonoVertexPhase = "Running"
 	MonoVertexPhaseFailed  MonoVertexPhase = "Failed"
+	MonoVertexPhasePausing MonoVertexPhase = "Pausing"
```
Comment (whynowy) [inline pkg/apis/numaflow/v1alpha1/mono_vertex_types.go:41]: No need to have Pausing.

File: pkg/reconciler/monovertex/controller.go [hunk 4]
```diff
@@ -132,16 +139,115 @@ func (mr *monoVertexReconciler) reconcile(ctx context.Context, monoVtx *dfv1.Mon
...
+	case dfv1.MonoVertexPhaseRunning, dfv1.MonoVertexPhaseUnknown:
+		return mr.resumeMonoVertex(ctx, mvtx)
+	default:
+		return false, fmt.Errorf("invalid desired phase")
+	}
+}
+
+func (mr *monoVertexReconciler) resumeMonoVertex(ctx context.Context, mvtx *dfv1.MonoVertex) (bool, error) {
+	// reset pause timestamp
+	if mvtx.GetAnnotations()[dfv1.KeyPauseTimestamp] != "" {
```
Comment (whynowy) [inline pkg/reconciler/monovertex/controller.go:187]: No need to have this annotation since no need to have graceful pausing.

**Comment by kohlisid** [comment]
Closing for conflicts. Will open a new PR

**Pull request closed without merging**